core/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/value/decorator/FromListVarEntityPropertyValueSelector.java

```diff
@@ -0,0 +1,121 @@
+import ai.timefold.solver.core.impl.domain.valuerange.descriptor.FromListVarEntityPropertyValueRangeDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.descriptor.GenuineVariableDescriptor;
+import ai.timefold.solver.core.impl.heuristic.selector.AbstractDemandEnabledSelector;
+import ai.timefold.solver.core.impl.heuristic.selector.value.EntityIndependentValueSelector;
+import ai.timefold.solver.core.impl.heuristic.selector.value.FromEntityPropertyValueSelector;
+import ai.timefold.solver.core.impl.phase.scope.AbstractPhaseScope;
+
+/**
+ * The value range for list variables requires the selector to be entity-independent,
+ * as it needs to fetch the entire list of values.
+ * Fetching the list of values is not a problem when the value range is located within the solution class,
+ * serving as a single source of truth.
+ * In cases where it is entity-dependent,
+ * the list of entities must be read to generate the corresponding list of values.
+ * <p>
+ * This selector adapts {@link FromEntityPropertyValueSelector} to behave like an entity-independent selector
+ * and meets the requirement to retrieve the complete list of values.
+ * 
+ * @param <Solution_> the solution type
+ */
+public final class FromListVarEntityPropertyValueSelector<Solution_> extends AbstractDemandEnabledSelector<Solution_>
+        implements EntityIndependentValueSelector<Solution_> {
+
+    private final FromEntityPropertyValueSelector<Solution_> childValueSelector;
+    private final boolean randomSelection;
+    private final FromListVarEntityPropertyValueRangeDescriptor<Solution_> valueRangeDescriptor;
+    private CountableValueRange<Object> cachedValueRange = null;
+
+    public FromListVarEntityPropertyValueSelector(FromEntityPropertyValueSelector<Solution_> childValueSelector,
+            boolean randomSelection) {
+        if (!(childValueSelector.getVariableDescriptor()
+                .getValueRangeDescriptor() instanceof FromListVarEntityPropertyValueRangeDescriptor<Solution_> descriptor)) {
+            throw new IllegalArgumentException(
+                    "The value range descriptor must be instance of %s."
+                            .formatted(FromListVarEntityPropertyValueRangeDescriptor.class.getSimpleName()));
+        }
+        this.childValueSelector = childValueSelector;
+        this.randomSelection = randomSelection;
+        this.valueRangeDescriptor = descriptor;
+    }
+
+    // ************************************************************************
+    // Life-cycle methods
+    // ************************************************************************
+    @Override
+    public void phaseStarted(AbstractPhaseScope<Solution_> phaseScope) {
+        super.phaseStarted(phaseScope);
+        this.childValueSelector.phaseStarted(phaseScope);
+        var scoreDirector = phaseScope.getScoreDirector();
+        cachedValueRange = valueRangeDescriptor.extractValueRange(scoreDirector.getWorkingSolution(), null);
```

Comment:
This can be on solver started.

Reply:
Good point. I wonder why `FromSolutionPropertyValueSelector` doesn't execute its logic at the solver start event.